[sw/dif_keymgr] Key Manager DIF Library Header

alphan opened the pull request:
This change adds the header of the key manager DIF library that defines its public interface.

Relevant docs:
* [Key Manager HWIP Technical Specification](https://docs.opentitan.org/hw/ip/keymgr/doc)
* [Identities and Root Keys](https://docs.opentitan.org/doc/security/specs/identities_and_root_keys/)

Signed-off-by: Alphan Ulusoy <alphan@google.com>

mcy commented on sw/device/lib/dif/dif_keymgr.c:
```diff
@@ -0,0 +1,9 @@
+
+#include "keymgr_regs.h"  // Generated.
+
+#error "Key Manager DIF library is not implemented yet."
```
I would drop the #error, since it makes it so this library can't be built. I think it's fine to just have an empty file so that the header is built and checked.

alphan replied:
This is based on @lenary's comment [here](https://github.com/lowRISC/opentitan/pull/3142#discussion_r493444840). I don't have a strong preference either way.

mcy commented on sw/device/lib/dif/dif_keymgr.h:
```diff
@@ -0,0 +1,648 @@
+  /**
+   * The call succeeded.
+   */
+  kDifKeymgrOK = 0,
```
Nit: OK -> Ok

mcy commented on sw/device/lib/dif/dif_keymgr.h:
```diff
@@ -0,0 +1,648 @@
+  /**
+   * \internal Last key manager interrupt.
+   */
+  kDifPwrmgrIrqLast = kDifKeymgrError,
```
Pwrmgr? =P

I would drop this and add it only if you wind up needing it in the impl.

mcy commented on sw/device/lib/dif/dif_keymgr.h:
```diff
@@ -0,0 +1,648 @@
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_keymgr_result_t dif_keymgr_initialize_state(const dif_keymgr_t *keymgr);
```
Thoughts: Is "initialize" a good word here? It feels confusable with "init". I don't know what the best word is, but I think something like "prime" or "seed" might be more appropriate.

alphan replied:
I had the same concern but thought prime/seed might be slightly inaccurate because this value is not used for computing the secret value of the following states.

This function can also be dropped but I am inclined to keep it until the HW design is finalized. Leaving this open for others to share their thoughts.

mcy commented on sw/device/lib/dif/dif_keymgr.h:
```diff
@@ -0,0 +1,648 @@
+/**
+ * Parameters for advancing the state of key manager.
+ */
+typedef struct dir_keymgr_advance_state_params {
+  /**
+   * This value is used by key manager to derive secret values and can be either
+   * a value that represents the contents of a boot stage, e.g. a truncated
+   * hash, or a tag. If it is a hash, changes in a boot stage will change the
+   * secret value, and consequently the versioned keys and identity seeds
+   * generated at subsequent boot stages. If it is a tag, those secret values,
+   * versioned keys, and identity seeds will be preserved across updates of the
+   * boot stage as long as the tag remains the same.
+   */
+  uint32_t sw_binding[DIF_KEYMGR_SW_BINDING_WORDS];
+  /**
+   * Maximum allowed version for keys generated at the next state.
+   */
+  uint32_t max_key_version;
+  /**
+   * Expected next state of the key manager. This value is used only for
+   * error checking.
+   */
+  dif_keymgr_state_t expected_next_state;
+} dif_keymgr_advance_state_params_t;
```
Do you think this should use the `transaction_t` naming style?

alphan replied:
This is not a parameter of advance_state anymore.

mcy commented on sw/device/lib/dif/dif_keymgr.h:
```diff
@@ -0,0 +1,648 @@
+extern "C" {
+#endif  // __cplusplus
+
+/**
+ * Size definitions.
+ */
+#define DIF_KEYMGR_SW_BINDING_WORDS 4
+#define DIF_KEYMGR_SALT_WORDS 4
+#define DIF_KEYMGR_NUM_SHARES 2
+#define DIF_KEYMGR_SHARE_WORDS 8
```
Optional: to avoid what are technically unnecessary macros, I think it would be ok to write something like
```c
enum {
  kDifKeymgrSwBindingWords = 4,
};
```

Me and @lenary have thought about doing this to avoid unnecessary macro names I don't think we should block on it.

mcy commented on sw/device/lib/dif/dif_keymgr.h:
```diff
@@ -0,0 +1,648 @@
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_keymgr_lockable_result_t dif_keymgr_disable(const dif_keymgr_t *keymgr);
```
This should be called `dif_keymgr_lock()`, I think?

(With an `is_locked()` companion.)

alphan replied:
I considered this option too, but chose this name to mostly align with the tech spec. Also, since we already have a getter for the status, I thought is_locked would be redundant.

mcy commented on sw/device/lib/dif/dif_keymgr.h:
```diff
@@ -0,0 +1,648 @@
+  /**
+   * Software invoked an invalid operation.
+   */
+  kDifKeymgrStatusCodeInvalidOperation = (1 << 0),
```
Note; the parens are unnecessary.

I would consider making these just regular bit indices, so that they can be used with `bitfield_bit32_*()`. `otp_ctrl` has this kind of API (which makes the indices usable as array indices, as well). You don't *need* it for this API but it would be nice if all the DIFs were consistent on this front.

alphan replied:
I definitely agree with the consistency argument. I don't have a strong preference for either but IMHO defining these as masks make it a bit more obvious how these are supposed to be used but OTOH eliminates the possibility of using `bitfield_bit32_*()` by callers, which may hurt readability.

mcy commented on sw/device/lib/dif/dif_keymgr.h:
```diff
@@ -0,0 +1,648 @@
+/**
+ * Key manager status.
+ */
+typedef struct dif_keymgr_status {
+  /**
+   * Current state of key manager.
+   */
+  dif_keymgr_state_t state;
+  /**
+   * Operational status of key manager.
+   *
+   * The following snippet can be used to check if key manager is idle:
+   *
+   *   bool is_idle = (status.status_codes & kDifKeymgrStatusCodeIdle);
+   *
+   * The following snippet can be used to check if key manager is idle and
+   * error-free:
+   *
+   *   bool is_idle_and_ok = (status.status_codes == kDifKeymgrStatusCodeIdle);
+   */
+  uint32_t status_codes;
+} dif_keymgr_status_t;
```
Do these *need* to be queried together?

alphan replied:
Not really, but currently HW handles initialization in a different way for the sake of efficiency (not as an operation but as a separate command and reports its status in WORKING_STATE.STATE instead of OP_STATUS.STATUS). I tried to abstract this away (HW may change to do the same). Getting the status would require reading both registers anyway so I thought it would be convenient to keep everything related in a single struct.

mcy commented on sw/device/lib/dif/dif_keymgr.h:
```diff
@@ -0,0 +1,648 @@
+   * versioned keys, and identity seeds will be preserved across updates of the
+   * boot stage as long as the tag remains the same.
+   */
+  uint32_t sw_binding[DIF_KEYMGR_SW_BINDING_WORDS];
```
Is there a particular reason this isn't a byte array? I assume the alignment isn't tremendously important.

mcy commented on sw/device/lib/dif/meson.build:
```diff
@@ -174,6 +174,21 @@ sw_lib_dif_pwrmgr = declare_dependency(
   )
 )
 
+# Key Manager DIF library
+# sw_lib_dif_keymgr = declare_dependency(
+#   link_with: static_library(
+#     'sw_lib_dif_keymgr',
+#     sources: [
+#       hw_ip_keymgr_reg_h,
+#       'dif_keymgr.c',
+#     ],
+#     dependencies: [
+#       sw_lib_mmio,
+#       sw_lib_bitfield,
+#     ],
+#   )
+# )
```
Can you uncomment this so the header gets compiled as part of CI?

alphan replied:
Related to [this](https://github.com/lowRISC/opentitan/pull/4256#discussion_r526987407).

mcy commented on sw/device/lib/dif/dif_keymgr.h:
```diff
@@ -0,0 +1,648 @@
+  /**
+   * Key manager issued invalid data to KMAC interface.
+   */
+  kDifKeymgrStatusInvalidKMACInput = (1 << 2),
```
Nit: I think we're pretty consistent in not all-caps-ing acronyms in constants (so KMAC -> Kmac). I the style guide actually mandates this spelling.

mcy reviewed (commented):
Hi Alphan, I've done a first-pass review on the code so far. LMK what you think. =)

alphan commented:
Thanks for the quick review @mcy! It looks like I accidentally created this as a ready-for-review PR from github cli, apologies for the noise :) I'll ping you once I hear from HW folks and update the PR.

mcy commented:
Haha, sounds good. =)

lenary commented:
Garret and I have scheduled a Design Review of this for Monday, 23rd November 2020.

alphan commented on sw/device/lib/dif/dif_keymgr.h:
```diff
@@ -0,0 +1,683 @@
+  uint32_t max_key_version;
+} dif_keymgr_state_params_t;
+
+/**
+ * Sets parameters for a key manager state.
+ *
+ * This function sets and locks the max version value and the binding value that
+ * is used to compute the secret value for the operational states of key
+ * manager, i.e. CreatorRootKey, OwnerIntermediateKey, and OwnerRootKey.
+ *
+ * @param keymgr A key manager handle.
+ * @param state A key manager state.
+ * @param params State parameters.
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_keymgr_lockable_result_t dif_keymgr_set_state_params(
+    const dif_keymgr_t *keymgr, dif_keymgr_state_t state,
+    dif_keymgr_state_params_t params);
```
I initially had this as a part of `dif_keymgr_advance_state()` but based on our discussions w/ @tjaychen and @moidx we will probably need a way to set and lock state inputs before advancing state due to lack of entropy. `dif_keymgr_state_params_t` does not hold the parameters of every state because every stage in the secure boot chain will set the parameters of the next state.

Regarding locking: There is no need to postpone locking but we may want to handle it separately by defining `*_lock` and `*_is_locked` depending on how we want to handle hardening.

alphan replied:
I decided to remove this function because there is a single register for the binding value now.

alphan commented on sw/device/lib/dif/dif_keymgr.h:
```diff
@@ -0,0 +1,683 @@
+   * The register that needs to be written to is locked.
+   */
+  kDifKeyMgrLockableLocked,
+} dif_keymgr_lockable_result_t;
```
I'm aware that this is currently missing a `Busy` constant. I need to confirm the behavior w/ the HW team.

alphan replied:
I don't think we need a separate enum for busy.

alphan commented on sw/device/lib/dif/dif_keymgr.h:
```diff
@@ -0,0 +1,683 @@
+/**
+ * Valid lengths for the `binding_value` array in `dif_keymgr_state_params_t`.
+ */
+typedef enum dif_keymgr_binding_value_len {
+  /**
+   * `binding_value` length when transitioning to `OwnerIntermediateKey` and
+   * `OwnerRootKey` states.
+   */
+  kDifKeymgrBindingValueLenTag = 4,
+  /**
+   * `binding_value` length when transitioning to `CreatorRootKey` state.
+   */
+  kDifKeymgrBindingValueLenRomExtHash = 8,
+} dif_keymgr_binding_value_len_t;
```
`RomExtSecurityDescriptor` is currently a 128b value in the tech spec but based on our discussions it is expected to be 256b. We may not need these constants if the software binding value ends up using the same size. Using a union in `dif_keymgr_state_params_t ` instead of this is also another option.

alphan replied:
`RomExtSecurityDescriptor`is now implemented through software binding (#4486).

gkelly commented:
Just a gentle reminder, reviews before header reviews should probably be avoided in order to reduce the toil for the header author.

lenary commented on sw/device/lib/dif/dif_keymgr.md:
```diff
@@ -0,0 +1,71 @@
+Implementation | [DIF_EXISTS][]       | Not Started |
+Implementation | [DIF_USED_IN_TREE][] | Not Started |
+Tests          | [DIF_TEST_UNIT][]    | Not Started |
+Tests          | [DIF_TEST_SMOKE][]  | Not Started |
+Review         | Reviewer(s)          | Not Started |
+Review         | Signoff date         | Not Started |
+
+[DIF_EXISTS]:       {{< relref "/doc/project/checklist.md#dif-exists" >}}
+[DIF_USED_IN_TREE]: {{< relref "/doc/project/checklist.md#dif-used-in-tree" >}}
+[DIF_TEST_UNIT]:    {{< relref "/doc/project/checklist.md#dif-test-unit" >}}
+[DIF_TEST_SMOKE]:  {{< relref "/doc/project/checklist.md#dif-test-smoke" >}}
```
Nit: Can you add spaces in the `smoke` lines? I see you did a search/replace for sanity, sadly they have different lengths.

gkelly commented on sw/device/lib/dif/dif_keymgr.h:
```diff
@@ -0,0 +1,683 @@
+   * Secret value of key manager is initialized with random values generated by
+   * the entropy source.
+   */
+  kDifKeymgrStateInitialized,
```
@tjaychen For the `Init_wip` state: why does it exist, and exposed to software? Is there some special treatment of the device we need to do while it's active? The other state transitions are asynchronous and aren't visible to software, so I'm curious about the difference.

alphan replied:
`Init_wip` was removed in #4467.

gkelly reviewed (commented):
Thanks for writing this up, @alphan! @lenary and I did a DIF review of this change. We noticed that there's some strangeness around the hardware interface, an we've tagged the HWEs on those questions. We'd like to see some separation in the status/state accessors so that a user can query one at a time. We'd like to hold off on the implementation until we've seen answers to the hardware concerns we raised.

Thanks!

alphan commented on sw/device/lib/dif/dif_keymgr.h:
```diff
@@ -0,0 +1,683 @@
+/**
+ * Key manager interrupts.
+ */
+typedef enum dif_keymgr_irq {
+  /**
+   * Operation was completed.
+   *
+   * Note: This interrupt is triggered regardless of the outcome of the
+   * operation. An operation is successful if only `kDifKeymgrIrqDone` is
+   * triggered and unsuccessful if both `kDifKeymgrIrqDone` and
+   * `kDifKeymgrIrqError` are triggered.
+   */
+  kDifKeymgrIrqDone,
+  /**
+   * Key manager encountered an error while performing the operation.
+   *
+   * There are two categories of errors: hardware fault errors and software
+   * operation errors. Hardware fault errors are caused by invalid KMAC
+   * commands, states, or outputs and are not related to software. Software
+   * operation errors are caused by invalid inputs from software. Clients can
+   * use `dif_keymgr_get_status()` to determine the type of error that occurred.
+   */
+  kDifKeymgrIrqError,
+  /**
+   * \internal Last key manager interrupt.
+   */
+  kDifKeymgrIrqLast = kDifKeymgrError,
+} dif_keymgr_irq_t;
```
Error IRQ is removed in #4513.

alphan commented on sw/device/lib/dif/dif_keymgr.h:
```diff
@@ -0,0 +1,683 @@
+   *
+   *   bool is_idle_and_ok = (status.status_codes == kDifKeymgrStatusCodeIdle);
+   */
+  uint32_t status_codes;
```
Yes, good catch! But we don't have this struct anymore (after defining individual getters for the state and status codes).

alphan commented on sw/device/lib/dif/dif_keymgr.h:
```diff
@@ -0,0 +1,683 @@
+ *
+ * This is an asynchronous function because key manager state transitions
+ * involve KMAC operations that can take some time to complete. Clients must
+ * check the status of key manager using `dif_keymgr_get_status()` before
```
IIUC, `OP_STATUS` (rw1c) will be `WIP` during an operation even if the software clears it. Once the operation ends, the software will read `DONE_SUCCESS` or `DONE_ERROR` and clearing those will reset `OP_STATUS` to `Idle`. @tjaychen, is this correct?

tjaychen commented:
yes that's right.  At least that's how it was intended, can't say there
aren't bugs of course :)

On Wed, Jan 6, 2021 at 1:16 PM Alphan Ulusoy <notifications@github.com>
wrote:

> *@alphan* commented on this pull request.
> ------------------------------
>
> In sw/device/lib/dif/dif_keymgr.h
> <https://github.com/lowRISC/opentitan/pull/4256#discussion_r552964871>:
>
> > +    const dif_keymgr_t *keymgr, dif_keymgr_state_t state,
> +    dif_keymgr_state_params_t params);
> +
> +/**
> + * Advances key manager state.
> + *
> + * This function initiates a transition to the next state, i.e. Initialized ->
> + * CreatorRootKey -> OwnerIntermediateKey -> OwnerRootKey -> Disabled. State
> + * transitions are irreversible during a power cycle. Key manager must be
> + * initialized using `dif_keymgr_initialize_state()` and parameters for the next
> + * state must be set using `dif_keymgr_set_state_params()`  prior to calling
> + * this function.
> + *
> + * This is an asynchronous function because key manager state transitions
> + * involve KMAC operations that can take some time to complete. Clients must
> + * check the status of key manager using `dif_keymgr_get_status()` before
>
> IIUC, OP_STATUS (rw1c) will be WIP during an operation even if the
> software clears it. Once the operation ends, the software will read
> DONE_SUCCESS or DONE_ERROR and clearing those will reset OP_STATUS to Idle.
> @tjaychen <https://github.com/tjaychen>, is this correct?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/lowRISC/opentitan/pull/4256#discussion_r552964871>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAH2RSW22FAC64WMY37FMMTSYTHKJANCNFSM4T3NQIGA>
> .
>

alphan commented on sw/device/lib/dif/dif_keymgr.h:
```diff
@@ -0,0 +1,683 @@
+ *
+ * See also: `dif_keymgr_status_codes_t`.
+ */
+typedef enum dif_keymgr_status_code {
```
Yes, you're right. It looks like I named the error values looking at the description in the spec instead of the register table. `kDifKeymgrStatusInvalidInput` should have been `kDifKeymgrStatusInvalidKmacCommand`.

Regarding translation: Yes, there is a one-to-one mapping between the bits of the `ERROR_CODE` register and the first four bits of `dif_keymgr_status_codes_t`. The fifth bit is defined as `OP_STATUS != WIP`. Other values of `OP_STATUS` is used to determine if we need to check the `ERROR_CODE` register or not.

alphan commented:
Thanks for reviewing this @gkelly, @lenary, and @mcy. @tjaychen, thanks for your time and the improvements in the software interface! Rebased and pushed a new commit (just to highlight the differences, will be squashed before merging) with the following changes (mostly in response to the changes in HW):
- Addressed reviewers' comments.
- Added `dif_keymgr_clear_sideload_keys()`.
- Added `dif_keymgr_configure()` and `dif_keymgr_config_t` to support reseed interval.
- Removed error interrupt.
- Removed `dif_keymgr_initialize_state()`
- Removed `RomExtSecurityDesc` related definitions, since binding value is used for both now.
- Removed `dif_keymgr_set_state_params()` and moved next state parameters to `dif_keymgr_advance_state()`.
- Updated comments as needed.

moidx commented on sw/device/lib/dif/dif_keymgr.h:
```diff
@@ -0,0 +1,656 @@
+   * identity seeds will be preserved across updates of the boot stage as long
+   * as the tag remains the same.
+   */
+  uint8_t binding_value[16];
```
Sorry, late to the review party, would it be ok to list this as `uint32_t binding_value[4]`?

alphan replied:
Done.

moidx commented on sw/device/lib/dif/dif_keymgr.h:
```diff
@@ -0,0 +1,656 @@
+   * identity seeds will be preserved across updates of the boot stage as long
+   * as the tag remains the same.
+   */
+  uint8_t binding_value[16];
```
@alphan, @tjaychen 

If we are using `binding_value` as the `ROMExtDescriptor`,  would it be possible to make the binding value `256B` so that we can use hash values if needed?

alphan replied:
Done.

moidx commented on sw/device/lib/dif/dif_keymgr.h:
```diff
@@ -0,0 +1,656 @@
+  /**
+   * Salt value to use for key generation.
+   */
+  uint8_t salt[16];
```
Is it possible to make this `uint32_t salt[4]`?

alphan replied:
Done.

moidx commented on sw/device/lib/dif/dif_keymgr.h:
```diff
@@ -0,0 +1,656 @@
+ * Key manager outputs are in two-shares.
+ */
+typedef struct dif_keymgr_output {
+  uint8_t value[2][32];
```
`uint32_t value[2][8]`

alphan replied:
Done

moidx commented on sw/device/lib/dif/dif_keymgr.h:
```diff
@@ -0,0 +1,656 @@
+  /**
+   * Salt value to use for key generation.
+   */
+  uint8_t salt[16];
```
@tjaychen, @alphan 

Since we are combining ID and salt (I assume), is it possible to make this 256B?

alphan replied:
Done.

moidx reviewed (commented):
Added a few comments.

tjaychen commented:
hey all, regarding increasing the width of sw_binding and salt to `256b`, that is fine with me.
But can we clear it with @moidx to make sure everything is inline?

Then we should first push a separate PR to update the identities security doc to make sure there is no confusion later.

mateenfaisal commented:
What is blocking this PR? pinging reviewers @silvestrst, @mcy....

silvestrst commented:
> What is blocking this PR? pinging reviewers @silvestrst, @mcy, @moidx, @gkelly

I haven't reviewed this change, can do next week, otherwise if others think it is ready, they should approve and merge it.

gkelly reviewed (approved):
Please follow update to @moidx's new types before submitting. Thanks @alphan!

alphan commented:
> hey all, regarding increasing the width of sw_binding and salt to `256b`, that is fine with me.
> But can we clear it with @moidx to make sure everything is inline?
> 
> Then we should first push a separate PR to update the identities security doc to make sure there is no confusion later.

I agree, created #5193 and #5194 to track doc and rtl changes.
